spec/system/admin/order_cycles/simple_spec.rb

```diff
@@ -399,13 +399,37 @@
           oc.update prefers_product_selection_from_coordinator_inventory_only: false
         end
 
-        it "shows a warning when going to 'outgoing products' tab" do
-          visit edit_admin_order_cycle_path(oc)
-          click_link "Outgoing Products"
-          within "tr.distributor-#{distributor_managed.id}" do
-            page.find("td.products").click
+        shared_examples "inventory warning" do
+          |inventory_visible, inventory_only, it_description, expect_message|
+          before do
+            # hides/displays variant within coordinator's inventory
+            inventory_item_v1.update!(visible: inventory_visible)
+            # changes coordinator's inventory preferences
+            supplier_managed.update preferred_product_selection_from_inventory_only: inventory_only
+          end
+
+          it "#{it_description} a warning when going to 'outgoing products' tab" do
+            visit edit_admin_order_cycle_path(oc)
+            click_link "Outgoing Products"
+            within "tr.distributor-#{distributor_managed.id}" do
+              page.find("td.products").click
+            end
+
+            # we need this assertion here to assure there is enough time to
+            # toggle the variant box and evaluate the following assertion
+            expect(page).to have_content product.name.upcase
+
+            # iterates between true / false, depending on the test case
+            expectation = expect_message ? :to : :not_to
+            expect(page).public_send(expectation,
+                                     have_content(%(No variant available for this product
+                (hidden via inventory settings)).squish))
           end
-          expect(page).to have_content "(Some variants might be hidden via inventory settings)"
+        end
+
+        it_behaves_like "inventory warning", false, true, "shows", true
+        it_behaves_like "inventory warning", false, false, "does not show", false do
```

Comment:
It's a bit hard to read I wonder if we could refactor this to use a proc with named argument , ie something like :
```ruby
def assert_inventory_warning(expectation, inventory_visible:, inventory_only:)
  ...
  case expectation
    when "show"
      expect_message= :to 
    when "does not show"   
      expect_message= :to 
     
  expect(page).public_send(expectation,
                                     have_content(%(No variant available for this product
                                                           (hidden via inventory settings)).squish))
end

it "shows a warning when going to 'outgoing products' tab" do
   assert_inventory_warming("shows", inventory_visible: false, inventory_only: true)
end
```

Reply:
I find keeping specs DRY is problematic! I think sometimes we try too hard, and it's ok to have some repetition. 

Sometimes shared_examples makes things harder to read (especially when trying to see where a failure occurred), so I like Gaetan's suggestion of using a method.

But also, this works and I think is fine as it is 👍

Reply:
> I find keeping specs DRY is problematic! I think sometimes we try too hard, and it's ok to have some repetition.

Yes, agree @dacook - it would be ok I guess to have two similar cases... My original idea was to have shared_examples of the 4 combinations (invoice settings (true/false) `x` variant hidden (true / false)). Then I thought it might be an overkill, but thinking about it now, it makes sense, also because there are two types of "invoice hidden variant" messages, which comes from scenarios where no variants are hidden, some are hidden, all are hidden + feature is disabled.

There are some possibilities left to cover, worth additional thought on this. Thanks for the suggestion @rioug , I'll try to make it work in a follow-up for this.